### DISCUSSION TITLE
Balloon in Immersive Mode

### DISCUSSION THREAD

**Discussion by JoshMcRose** [skydoves/Balloon#295] [Q&A]
Hi @skydoves,

I'm a bit new to Android development, so maybe I'm missing something, but is it possible to build a balloon in Immersive Mode without removing Immersive Mode?

Thanks for sharing this cool library, by the way!

**Comment by skydoves** [comment]
Hey @ubertrombone,
Could you expand on this?

**Reply by JoshMcRose** [reply]
Sure! So my app is full screen, but when the balloon shows up, the system task bar and navigation appear:
![Screenshot_20211231-201017_Shattered.jpg](https://user-images.githubusercontent.com/66621641/147835247-452c63b6-8e3f-4e72-93b3-c2fde463d223.jpg)

For Dialog Fragments I solve this problem by putting the following code in onCreateView: 

dialog?.window?.setFlags(WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE, WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE)
@Suppress("DEPRECATION")
dialog?.window?.decorView?.systemUiVisibility = requireActivity().window.decorView.systemUiVisibility

dialog?.setOnShowListener {
dialog?.window?.clearFlags(WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE)

val wm = requireActivity().getSystemService(Context.WINDOW_SERVICE) as WindowManager
wm.updateViewLayout(dialog?.window?.decorView, dialog?.window?.attributes)
}

I hope that's helpful; thanks!

**Reply by skydoves** [reply]
Hey @ubertrombone,
I used the below codes for testing the immersive mode on the demo project following [this reference](https://developer.android.com/training/system-ui/immersive).

```kotlin
private fun hideSystemBars() {
    val windowInsetsController =
      ViewCompat.getWindowInsetsController(window.decorView) ?: return
    // Configure the behavior of the hidden system bars
    windowInsetsController.systemBarsBehavior =
      WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
    // Hide both the status bar and the navigation bar
    windowInsetsController.hide(WindowInsetsCompat.Type.systemBars())
  }
```

And it looks work for me like the screenshots below:

<p>
<img src="https://user-images.githubusercontent.com/24237865/147844734-1222020a-4b6d-498f-bd0c-51269be3547f.jpg" width="33%" />
<img src="https://user-images.githubusercontent.com/24237865/147844737-4a2ec0a2-2c90-4963-b105-d7e4f8254578.jpg" width="33%" />
</p>

Please correct me if I missing something. Thanks!

**Comment by JoshMcRose** [comment]
Hey @skydoves,

Happy New Year!

I do something similar in my activity (which I also call in my fragments) which puts my app in Immersive mode, but when the balloon shows up, it shows the system bars (like in my last picture). When I dismiss the balloon I call the function from my activity that returns the immersive mode, but that hasn't solved the issue when the balloon is shown. 

This is the code in my [activity](https://github.com/ubertrombone/shattered/blob/main/app/src/main/java/com/example/shattered/MainActivity.kt): 
fun fullScreen() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val insetsController = window.decorView.windowInsetsController ?: return
            insetsController.systemBarsBehavior = WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            insetsController.hide(WindowInsets.Type.systemBars())
        } else {
            @Suppress("DEPRECATION")
            window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                    or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_FULLSCREEN)
        }
    }

And this is how I'm building my balloon in the [fragment](https://github.com/ubertrombone/shattered/blob/main/app/src/main/java/com/example/shattered/game/GameFragment.kt):
val balloon = createBalloon(requireContext()) {
            setWidthRatio(1.0f)
            setHeight(BalloonSizeSpec.WRAP)
            setText("Edit your profile here!")
            setTextColorResource(R.color.white)
            setTextSize(15f)
            setIconDrawableResource(R.drawable.home_icon)
            setArrowPositionRules(ArrowPositionRules.ALIGN_ANCHOR)
            setArrowSize(10)
            setArrowPosition(0.5f)
            setPadding(12)
            setCornerRadius(8f)
            setBackgroundColorResource(R.color.royal_blue)
            setBalloonAnimation(BalloonAnimation.ELASTIC)
            setLifecycleOwner(viewLifecycleOwner)
            setOnBalloonClickListener { Toast.makeText(context, "clicked", Toast.LENGTH_SHORT).show() }
            setOnBalloonDismissListener { Toast.makeText(context, "dismissed", Toast.LENGTH_SHORT).show() }
            build()
        }
        binding?.helpButton?.showAlignBottom(balloon)
        (activity as MainActivity).fullScreen()

**Reply by rodo117** [reply]
Use this property  .setFocusable(false) in your balloon object , this is going to solve your problem
